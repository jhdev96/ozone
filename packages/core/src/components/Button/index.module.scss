@import '../../scss/root';
@import '../../scss/mixins/button/button.scss';
@import '../../scss/mixins/button/variables.scss';
                                  

// For styling `flat` buttons
$flatMap: (
  primary: (
    color: #fff,
    bg-color: var(--blue-3)
  ),
  secondary: (
    color: #fff,
    bg-color: var(--red-3)
  ),
  warning: (
    color: var(--black-core),
    bg-color: var(--yellow-3)
  )
);

/* Major button classes - primary, secondary, warning, basic */                                   
.primary {
  @include button;

  color: #fff;
  background-color: var(--blue-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;

    color: var(--blue-core);
    border: 1px solid var(--blue-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--red-core) auto 1px !important;
  }
}

.secondary {
  @include button;

  color: #fff;
  background-image: $secondary-gradient;
  background-repeat: repeat-x;
  border-color: $button-bc1 $button-bc2 $button-bc3 !important;
  box-shadow: inset 0 1px 0 $primary-shade1, 
                    0 1px 2px $primary-shade2;
  background-color: var(--red-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;
    color: var(--red-core);
    border: 1px solid var(--red-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--blue-core) auto 1px !important;
  }
}

.warning {
  @include button;

  color: var(--black-core);
  background-color: var(--yellow-core);
  border-radius: 4px;

  &:focus {
    outline: var(--black-core) auto 1px !important;
  }
}

.basic {
  @include button;

  border-radius: 3px;

  &:focus {
    outline: var(--gray-4) auto 1px !important;
  }
}

/* The following styles serve as modifiers 
 * for the major button classes 
*/
.flat {
  background-image: none !important;
  background-repeat: no-repeat;
  border: 0;
  border-radius: 0 !important;
  box-shadow: none !important;
  text-shadow: none !important;

  @each $class, $rule in $flatMap {
    &.#{$class}:not(.disabled):not(.loading) {
      &:hover {
        color: map-get($rule, 'color');
        background-color: map-get($rule, 'bg-color');
      }
    }
  }
}

.fill { width: 100%; }

.loading {
  padding: 8px 20px 5px 20px;
  max-height: 35px;
  opacity: 0.90;
  cursor: default;

  &:focus {outline: none !important;}
}

.disabled {
  opacity: 0.70;
  cursor: default;
}

.small {
  min-width: fit-content;
  padding: 8px !important;
}

.medium { min-width: 40px; }

.large {
  min-width: 70px;
  max-height: 70px !important;
  padding: 15px 20px !important;
  font-size: 1.2em !important;
}

.light {
  &.basic {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }
}

.dark {
  &.basic {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}