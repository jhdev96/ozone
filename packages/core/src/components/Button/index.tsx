import React from 'react';
import classNames from '../../utils/classNames';
import { 
  ButtonVariation,
  loaderTheme
} from './utils';
import LoadingDots from '../LoadingDots';
import styles from './index.module.scss';


export interface ButtonProps {
  /** 
   * The main classes of buttons. 
   */
  variation?: ButtonVariation;
  /** 
   * Sets the HTML `type` attribute for the
   * button element.
   */ 
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"]
  /** 
   * Control the size of the button. Default is medium. 
   */
  size?: "small" | "medium" | "large";
  /** 
   * Give the button a flat texture, appropriate for forms. 
   */
  flat?: boolean;
  /** 
   * When true the button will no background; 
   * It will only have a border. 
   */
  outline?: boolean;
  /** 
   * When true, the button's width will be 100% of 
   * its container. 
   */ 
  fill?: boolean;
  /** 
   * Color the button based on the current theme. 
   */
  theme?: "light" | "dark";
  /** 
   * Make the button disabled. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates whether the button is loading some content. 
   */
  isLoading?: boolean;
  /** 
   * Sets the `className` attribute on the button element. 
   */ 
  className?: string;
  /** 
   * The function that handles click events. 
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers over the button.
   */
  onMouseEnter?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers away from the
   * button.
   */
  onMouseLeave?: React.MouseEventHandler<HTMLButtonElement>;
  /** 
   * A selector used for testing this component. 
   */
  dataTestId?: string;
  /** 
   * Content displayed inside the button. 
   */
  children?: React.ReactNode | string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({
    variation     = "primary",
    type          = "button",
    size          = "medium", 
    flat          = false,
    outline       = false,
    fill          = false, 
    theme         = "light",
    isDisabled    = false,
    isLoading     = false,
    className,
    onClick       = () => {},
    onMouseEnter  = () => {},
    onMouseLeave  = () => {},
    dataTestId,
    children}: ButtonProps,
    ref: React.Ref<HTMLButtonElement>): JSX.Element 
  {
    const childMarkup = isLoading 
      ? (
          <>
            <LoadingDots 
              dotSize={7} 
              position="center" 
              theme={loaderTheme(variation, theme, outline)} 
            />
            <span style={{opacity: '0'}}>{children}</span>
          </>
        )
      : children;

    return (
      <button
        type={type}
        className={classNames(
          className,
          styles[variation],
          styles[size],
          styles[theme],
          flat        && styles.flat,
          outline     && styles[`${variation}--outline`],
          fill        && styles.fill,
          isLoading   && styles.loading,
          isDisabled  && styles.disabled
        )}
        onClick={isDisabled ? () => {} : onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        disabled={isDisabled}
        aria-busy={isLoading}
        data-testid={dataTestId}
        ref={ref}
      >
        {childMarkup}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;