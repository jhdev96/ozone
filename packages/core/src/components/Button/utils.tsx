import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}