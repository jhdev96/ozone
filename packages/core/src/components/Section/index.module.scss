@import '../../scss/root';


.section {
  position: relative;
  display: block;
  width: fit-content;
  margin: 10px 0;
  padding: 20px 10px 35px 10px;
  border-radius: 6px;
  border-top: none !important;
  box-sizing: content-box;

  .title {
    position: relative;
    width: fit-content;
    margin: 0 6px;
  }
  
  .header {
    position: relative;
    display: inline-flex;
    margin-left: 10px;
    top: -27px;
    width: 100%;
    white-space: nowrap;
  }

  .body {
    position: relative;
    padding: 0 15px;
    box-sizing: border-box;
  }

  .topLeftBorder {
    border-radius: 6px 0 0 0;
    width: 20px;
    height: 10px;
    position: absolute;
    top: 0px;
    left: 0px;
  }
  
  .topRightBorder {
    position: relative;
    top: 7px;
    width: 100%;
    height: 10px;
    border-radius: 0 6px 0 0;
  }
}

// variation
.primary {
  border: 1px solid var(--blue-core);

  .title { color: var(--blue-core); }
  .topLeftBorder { border-top: 1px solid var(--blue-core); }
  .topRightBorder { border-top: 1px solid var(--blue-core); }
}

.basic {
  border: 1px solid var(--gray-4);

  .title { color: var(--black-2); }
  .topLeftBorder { border-top: 1px solid var(--gray-4); }
  .topRightBorder { border-top: 1px solid var(--gray-4); }
}

// Theme
.light {
  .body {
    color: var(--black-core);
  }
}

.dark {
  .body {
    color: var(--gray-1);
  }
}

.fullWidth { width: 100%; }