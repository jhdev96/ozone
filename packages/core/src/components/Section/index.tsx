import React from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


interface SectionProps {
  /** 
   * The title displayed at the top of 
   * the section. 
   */
  title?: string;
  /** 
   * The main section classes. 
   */
  variation?: "primary" | "basic";
  /** 
   * Change the color of the text inside the 
   * section body to match the theme. 
   */
  theme?: ThemeType;
  /** 
   * Indicates whether the section should occupy 
   * the full width of its container. 
   */
  fullWidth?: boolean;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The content rendered inside a section. 
   */
  children?: React.ReactNode;
}

export default function Section({
  title       = "Section title",
  variation   = "primary",
  theme       = "light",
  fullWidth   = true,
  dataTestId,
  children
}: SectionProps): JSX.Element 
{
  return (
    <section 
      className={classNames(
        styles.section,
        styles[variation],
        styles[theme], 
        fullWidth && styles.fullWidth,
      )}
      data-testid={dataTestId}
    >
      <div className={styles.topLeftBorder} />
      <div className={styles.header}>
        <div className={styles.title}>
          {title}
        </div>
        <div className={styles.topRightBorder} />
      </div>
      <div className={styles.body}>
        {children}
      </div>
    </section> 
  )
}

Section.displayName = 'Section';