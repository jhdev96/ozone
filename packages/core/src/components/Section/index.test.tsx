import React from "react";
import { render, screen } from "@testing-library/react";
import Section from ".";
import Button from "../Button";


describe('<Section />', () => {
  it('renders a primary section by default', () => {
    const {container} = render(
      <Section />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a section with a title', () => {
    render(
      <Section 
        dataTestId="test" 
        title="title" 
      />
    );
    
    expect(document.querySelector('.title'))
      .toHaveTextContent('title');
  });

  it('occupies the full width of its container', () => {
    render(
      <div>
        <Section dataTestId="test" fullWidth />
      </div>
    );

    expect(screen.getByTestId('test'))
      .toHaveClass('fullWidth');
  });

  it('renders children in the section body', () => {
    render(
      <Section dataTestId="test">
        <Button>text</Button>
      </Section>
    );

    expect(document.querySelector('.body'))
      .not.toBeEmptyDOMElement();
  });
});