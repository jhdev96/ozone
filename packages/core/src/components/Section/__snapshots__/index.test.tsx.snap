// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section /> renders a primary section by default 1`] = `
<div>
  <section
    class="section primary light fullWidth"
  >
    <div
      class="topLeftBorder"
    />
    <div
      class="header"
    >
      <div
        class="title"
      >
        Section title
      </div>
      <div
        class="topRightBorder"
      />
    </div>
    <div
      class="body"
    />
  </section>
</div>
`;
